### DISCUSSION TITLE
How to build italian ASR

### DISCUSSION THREAD

**Discussion by ghost** [flashlight/flashlight#475] [Q&A]
Hi!
I ran Flashlight for CPU backend with Docker as in https://github.com/facebookresearch/flashlight/tree/master/.docker#running-flashlight-with-docker
I have downloaded from https://github.com/facebookresearch/wav2letter/tree/master/recipes/mls the following files:
- tokens.txt
- train_lexicon.txt
- joint_lexicon.txt
- arch.txt
- am.bin
- mls_lm_italian.tar.gz
- mls_italian.tar.gz
but I did not find clear instructions on how to continue building the structure for the ASR for the Italian language.

Thanks for your help!

**Comment by dmzubr** [comment]
Hi!

Do you want to decode (inference) or train / finetuning with your extra audio data?

**Reply by ghost** [reply]
Hi, I want to decode (inference) audio data!

**Comment by tlikhomanenko** [comment]
Hey, 

If you want to retrain model from scratch then you need to follow these steps https://github.com/facebookresearch/wav2letter/tree/master/recipes/mls#usage to reproduce.

If you want just to reproduce decoding with pretrained model then follow decoding instructions here https://github.com/facebookresearch/wav2letter/tree/master/recipes/mls#decoding.

**Reply by ghost** [reply]
Thanks Tatiana for your availability,
I follow the tutorial "Run ASR Streaming Samples Quickly Using Docker" as in https://github.com/facebookresearch/wav2letter/wiki/Inference-Run-Examples#quickly-run-streaming-asr-examples-using-docker.
Specifically when I try "Interactive Streaming ASR Example" I can't get the transcription of the audio files with the related messages during processing as shown in the example:

Transcribing file: /home/audio/LibriSpeech/dev-clean/777/126732/777-126732-0011.flac.wav to: stdout
Creating LexiconDecoder instance.
#start (msec), end (msec), transcription
0.1000,
1000,2000, the
2000,3000, possessor of property had not
3000,4000, only to face the awakened
4000,5000, proletariat
5000,6000, but they had also
6000,7000, to fight amongst
7000,7450, themselves
#finish transcribing

while this message appears:

Transcribing file: /home/audio/LibriSpeech/dev-clean/777/126732/777-126732-0011.flac.wav to: stdout
Creating LexiconDecoder instance.
#start (msec), end (msec), transcription
Illegal instruction (core dumped) 

How can I fix the "Illegal instruction (core dumped)" message?

**Reply by tlikhomanenko** [reply]
cc @vineelpratap @avidov, they could help on this.

**Reply by vineelpratap** [reply]
Hi,
  The models trained doesn't support online decoding. However, you can run inference using the instructions provided in the recipes folder. 
  Please see the discussion in https://github.com/facebookresearch/wav2letter/issues/920#issuecomment-747726820 for more details .

**Reply by ghost** [reply]
Thanks Vineel for your suggestion,

I ran Beam-search Decoders command:

/root/flashlight/build/bin/asr/fl_asr_decode
--sclite=/root/flashlight/flashlight/app/asr/mls_it/results/lm_decoder/
--am=/root/flashlight/flashlight/app/asr/mls_it/am.bin
--lm=/root/flashlight/flashlight/app/asr/mls_it/mls_lm_italian/5-gram_lm.arpa
--tokens=/root/flashlight/flashlight/app/asr/mls_it/tokens.txt
--lexicon=/root/flashlight/flashlight/app/asr/mls_it/joint_lexicon.txt
--rundir=/root/flashlight/flashlight/app/asr
--emission_dir=/root/flashlight/flashlight/app/asr/mls_it/emission_dir/
--datadir=/root/flashlight/flashlight/app/asr/mls_it/audio/
--show
--showletters
--maxload=1
--nthread_decoder=2
--beamsize=100
--beamsizetoken=100
--beamthreshold=20
--lmweight=1
--wordscore=0
--eosscore=0
--silscore=0
--unkscore=0
--smearing=max
--lmtype=kenlm
--decodertype=wrd
--test=gr1sport_10s.lst

and this message appears:

terminate called after throwing an instance of 'std::runtime_error'
what(): failed to open file for reading: /root/flashlight/flashlight/app/asr/mls_it/emission_dir/gr1sport_10s.lst/gr1sport_10s.bin
*** Aborted at 1616431904 (unix time) try "date -d @1616431904" if you are using GNU date ***
PC: @ 0x7f9bcca9318b gsignal
*** SIGABRT (@0xdac) received by PID 3500 (TID 0x7f9bc7f0f980) from PID 3500; stack trace: ***
@ 0x7f9bd8225631 (unknown)
@ 0x7f9bd11093c0 (unknown)
@ 0x7f9bcca9318b gsignal
@ 0x7f9bcca72859 abort
@ 0x7f9bcce8b951 (unknown)
@ 0x7f9bcce9747c (unknown)
@ 0x7f9bcce974e7 std::terminate()
@ 0x7f9bcce9746f std::rethrow_exception()
@ 0x563a49162c34 std::future<>::get()
@ 0x563a490c43ab main
@ 0x7f9bcca740b3 __libc_start_main
@ 0x563a4914bf4e _start
Aborted (core dumped)

How can I fix the "failed to open file for reading: /root/flashlight/flashlight/app/asr/mls_it/emission_dir/gr1sport_10s.lst/gr1sport_10s.bin" message?

**Reply by vineelpratap** [reply]
Hi, can you keep `--emission_dir` path empty and try again.

**Reply by ghost** [reply]
Thanks Vineel,

using --emission_dir path empty it works!

**Comment by ghost** [comment]
I have a question about how to improve the transcription for the Italian language.
Is it sufficient to add only new words in the joint_lexicon.txt file or do I need to train a new acoustic model with a customized audio and transcription data set?
What's the correct way to improve transcription?

A further question is about the duration of the audio file to be transcribed, is there a limit on its duration?

**Reply by vineelpratap** [reply]
Hi, 
  What type of audio are you try to do inference on  - read speech or conversational ? Note that MLS is based on read speech (same as LibriSpeech) and it might not perform well on conversational speech. 
  I would try things in this order 
  - if you have more text data, add any new words to the provided lexicon and also train a new LM on MLS transcripts + new data with KenLM toolkit. Consider converting the `.arpa` file to `binary format` for  faster loading afterwards
  - if you happen to have additional ASR dataset in Italian which has similar style of speech, you can also finetune the provided MLS Italian model. Checkout this tutorial on how we finetune a model on English. You can try to adapt it for your use case - https://colab.research.google.com/github/facebookresearch/flashlight/blob/master/flashlight/app/asr/tutorial/notebooks/FinetuneCTC.ipynb

**Reply by ghost** [reply]
I have a 6GB data.txt and I have a few questions about training a new LM:

1) Is there a documentation/tutorial to follow for the language model training? Which dataset is needed? Is data.txt alone sufficient or are other files useful?

2) Are there any support commands to automatically create lexicon and vocab_counts files from data.txt or do they have to be created manually?

3) Can lexicon and lm files be created independently of the acoustic model?

**Reply by tlikhomanenko** [reply]
- You can use info from recipes how we trained ngram lms, for example https://github.com/facebookresearch/wav2letter/tree/master/recipes/lexicon_free/librispeech#reproduce-language-models-training--evaluation
- Nope, simply use python to do that, especially if your token set is letters it would be simple to do + your preparation depends how you train acoustic model (what token set)
- Yes, the only thing be care is to use lexicon spelling to be consistent with acoustic tokens. And your lm can be trained on any text.

**Reply by ghost** [reply]
Analyzing my text dataset I created the lexicon file considering the tokens file used for the acoustic model.
Analyzing the following link
https://github.com/facebookresearch/wav2letter/tree/master/recipes/lexicon_free/librispeech#reproduce-language-models-training--evaluation
I understand that this method refers to a Flashlight with Docker on the CUDA backend. 
I am using Flashlight with Docker on CPU backend, so I would like to know how to proceed to train a new language model? Are any more files needed to continue?

**Reply by tlikhomanenko** [reply]
For ngram you don't need cuda, so you can use cpu version in the same way.

**Comment by ghost** [comment]
A question about the duration of the audio file to be transcribed, is there a limit to the duration?

This question is due to the fact that after trying to transcribe a 300 second audio file both with Greedy Decoder, using only acoustic model predictions, and with Decoding, with embedded language model, the answer was blank after waiting for over 60 and 115 minutes respectively. Also, the transcript of a 120 second audio file was obtained after waiting 15 minutes.

So to get non-empty results and in faster times is it necessary to split the audio file to be transcribed into audio files of a certain duration?

**Reply by tlikhomanenko** [reply]
Are you using GPU to do inference with decoder?

**Reply by ghost** [reply]
I am using Flashlight with Docker on CPU backend.

**Reply by tlikhomanenko** [reply]
Because of the self attention the computation on each transfromer block is proportional to t*2, where t is input length. So will have very slow computations on the very long sequence. You need to chunk the input into the model.
